### PR TITLE
RP-3370 table auto creation implementation added

### PR DESCRIPTION
**Pull request by deepak-maheshwari-ext**

### CODE REVIEW

File: hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java [hunk 1]
```diff
@@ -27,16 +40,46 @@
...
+        LOG.error("Error while closing hbase admin {}", e.getMessage());
+      }
+    }
+  }
+
+  private void createTable(Configuration conf, HBaseAdmin admin) throws IOException {
+    String tableName = conf.get(HBASE_CUSTOM_TABLENAME);
+    List<String> listColumnFamily = Stream.of(conf.get(HBASE_CUSTOM_COLUMNFAMILY).split(","))
+        .collect(Collectors.toList());
+    if (admin.tableExists(tableName)) {
```
Comment (lav-mudgal-guavus) [inline hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java:74]: Need to handle scenario where table exists but column family doesn't.

File: hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java [hunk 2]
```diff
@@ -27,16 +40,46 @@
...
+      LOG.info("table {} already exists", tableName);
+    } else {
+      HTableDescriptor tableDesc = new HTableDescriptor(tableName);
+      listColumnFamily.forEach(columnFamily -> tableDesc.addFamily(new HColumnDescriptor(columnFamily)));
+      admin.createTable(tableDesc);
+      LOG.info("table has been created with name {}", tableName);
```
Comment (lav-mudgal-guavus) [inline hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java]: add column family in logs

File: hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java [hunk 3]
```diff
@@ -27,16 +40,46 @@
...
     ClassLoader originalClassLoader = Thread.currentThread().getContextClassLoader();
     Thread.currentThread().setContextClassLoader(getClass().getClassLoader());
+    HBaseAdmin admin = null;
     try {
       super.setConf(otherConf);
+      Configuration configuration = super.getConf();
+      admin = new HBaseAdmin(configuration);
+      createTable(configuration, admin);
+    } catch (IOException e) {
+      LOG.error("Error while creating hbase table {}", e.getMessage());
```
Comment (lav-mudgal-guavus) [inline hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java]: add stacktrace in error logs

**Review by lav-mudgal-guavus — CHANGES_REQUESTED** [review]
Please resolve review comments

File: hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java [hunk 4]
```diff
@@ -27,16 +40,74 @@
...
   @Override
   public void setConf(Configuration otherConf) {
     // To resolve CDAP-12731, set the current class's classloader to the thread's context classloader,
     // so that it gets picked up when the super.setConf() calls HBaseConfiguration.create(Configuration)
     ClassLoader originalClassLoader = Thread.currentThread().getContextClassLoader();
     Thread.currentThread().setContextClassLoader(getClass().getClassLoader());
+    HBaseAdmin admin = null;
     try {
       super.setConf(otherConf);
+      Configuration configuration = super.getConf();
```
Comment (zuxqoj) [inline hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java:53]: is it different from `otherConf` ??

File: hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java [hunk 5]
```diff
@@ -27,16 +40,74 @@
...
+      admin = new HBaseAdmin(configuration);
+      createOrUpdateTable(configuration, admin);
+    } catch (IOException e) {
+      LOG.error("Error while creating hbase table ", e);
```
Comment (zuxqoj) [inline hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java:57]: should throw exception

File: hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java [hunk 6]
```diff
@@ -27,16 +40,74 @@
...
     } finally {
       Thread.currentThread().setContextClassLoader(originalClassLoader);
+      try {
+        if (admin != null) {
+          admin.close();
+        }
+      } catch (IOException e) {
+        LOG.error("Error while closing hbase admin ", e);
```
Comment (zuxqoj) [inline hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java]: should be warn

File: hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java [hunk 7]
```diff
@@ -27,16 +40,74 @@
...
+          break;
+        }
+      }
+      // if not: add it
+      if (!exists) {
+        admin.addColumn(table.getTableName(), new HColumnDescriptor(family));
+        LOG.info("column family {} updated in table {}", family, table.getName().getNameAsString());
+      }
+    } catch (IOException e) {
+      LOG.error("error while adding column family", e);
```
Comment (zuxqoj) [inline hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java:108]: throw exception

File: hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java [hunk 8]
```diff
@@ -17,7 +17,20 @@
...
+import org.apache.hadoop.hbase.client.HTable;
 import org.apache.hadoop.hbase.mapreduce.TableOutputFormat;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+import static co.cask.hydrator.plugin.sink.HBaseSink.HBASE_CUSTOM_COLUMNFAMILY;
+import static co.cask.hydrator.plugin.sink.HBaseSink.HBASE_CUSTOM_TABLENAME;
```
Comment (zuxqoj) [inline hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java]: HBaseSink is consumer of HBaseTableOutputFormat class, reverse mapping should be avoided

File: hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java [hunk 9]
```diff
@@ -27,16 +40,74 @@
  */
 public class HBaseTableOutputFormat<KEY> extends TableOutputFormat<KEY> {
 
+  private static final Logger LOG = LoggerFactory.getLogger(HBaseTableOutputFormat.class);
+  public static final String DEFAULT_COL_FAMILY = "default";
```
Comment (zuxqoj) [inline hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java]: looks like you are not using it anywhere

**Review by zuxqoj — CHANGES_REQUESTED** [review]
In my opinion, we should change it to SparkSink based plugin

File: hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java (hunk 4 above)
Comment (deepak-maheshwari-ext) [reply hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java:53]: yes it have some more configs then other configs which is added by super class.

File: hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java (hunk 7 above)
Comment (deepak-maheshwari-ext) [reply hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java:108]: done

File: hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java (hunk 6 above)
Comment (deepak-maheshwari-ext) [reply hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java]: done

File: hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java (hunk 5 above)
Comment (deepak-maheshwari-ext) [reply hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java:57]: done

**Review by deepak-maheshwari-ext — COMMENTED** [review]
done

File: hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java [hunk 10]
```diff
@@ -27,16 +36,78 @@
...
     Thread.currentThread().setContextClassLoader(getClass().getClassLoader());
+    HBaseAdmin admin = null;
     try {
       super.setConf(otherConf);
+      Configuration configuration = super.getConf();
+      admin = new HBaseAdmin(configuration);
+      createOrUpdateTable(configuration, admin);
+    } catch (IOException e) {
+      LOG.error("Error while creating hbase table ", e);
+      throw new RuntimeException(e);
```
Comment (lav-mudgal-guavus) [inline hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java:58]: why we are converting it into RunTime ? let it be original exception

File: hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java [hunk 11]
```diff
@@ -27,16 +36,78 @@
...
+        }
+      }
+      // if not: add it
+      if (!exists) {
+        admin.addColumn(table.getTableName(), new HColumnDescriptor(family));
+        LOG.info("column family {} updated in table {}", family, table.getName().getNameAsString());
+      }
+    } catch (IOException e) {
+      LOG.error("error while adding column family", e);
+      throw new RuntimeException(e);
```
Comment (lav-mudgal-guavus) [inline hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java:109]: why Runtime ? let it be original exception

File: hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java (hunk 10 above)
Comment (deepak-maheshwari-ext) [reply hbase-plugins/src/main/java/co/cask/hydrator/plugin/sink/mapreduce/HBaseTableOutputFormat.java:58]: we can't throw directly IOExcption as it is checked exception.

**Review by lav-mudgal-guavus — APPROVED** [review]
LGTM

**Pull request merged**